### ISSUE TITLE
Accessing the worker scripts' scope

### ISSUE THREAD

**Issue by DoubleCouponDay** [AndersCan/typed-web-workers#14]
This library is an excellent addition to an angular workflow, and even gives access to libraries by downloading the `@next` branch. However, what I additionally need is to hold a web socket connection as a variable that lives between worker function runs. 

[From the mozilla docs:](https://developer.mozilla.org/en-US/docs/Web/API/Worker/onmessage) 

> Notice how in the main script, onmessage has to be called on myWorker, whereas inside the worker script you just need onmessage because the worker is effectively the global scope (DedicatedWorkerGlobalScope).

Would it be possible to add such functionality to the library? I'd imagine a third callback which initializes  variables attached to the script scope in a generic fashion would suit nicely.

**Comment by AndersCan** [comment]
> This library is an excellent addition to an angular workflow

Thanks :) 

> Would it be possible to add such functionality to the library? I'd imagine a third callback which initializes variables attached to the script scope in a generic fashion would suit nicely.

If I've understood correctly you want to create a web worker that has some internal state? State that is created in the worker scope during creation or something similar. I've thought about this issue before, but not really come up with any good solution.

I'll take a look again this weekend and get back to you.

**Comment by DoubleCouponDay** [comment]
Thank you!

**Comment by AndersCan** [comment]
Needed to spend some time fixing the test setup so this took a bit longer to get started on.

The `workerFunction` now takes two extra functions, getState and setState. So something like this is possible: 
```javascript
const workerFunction = (
  input,
  cb,
  getState,
  setState
) => {
  setState(input)
  cb(getState())
}
```
I'll have to make some more changes before I can get the `state` variable typed, but I'll work more on that later. In the meantime you can test this by running `npm install typed-web-workers@beta`. As it has been so long since my last `next` release, I've published this under `beta` just in case :)

**Comment by DoubleCouponDay** [comment]
Very cool. keep up the good work.

**Comment by AndersCan** [comment]
Okay. Holiday is over and I've had some time to work on this issue again :)

Biggest change I've now made is to `workerFunction` and that it now accepts an object with named parameters. As you can see below:

```javascript
const workerWithState = createWorker<
  number, // input type
  number[], // output type
  number[] // state type
>({
  workerFunction: ({
    input,
    cb,
    setState,
    getState
  }) => {
    setState(getState() || [])
    setState([...getState(), input])
    cb(getState())
  },
  onMessage: output => {
    console.log(output)
  }
})
workerWithState.postMessage(1) // will log: [1]
workerWithState.postMessage(2) // will log: [1,2]
workerWithState.postMessage(3) // will log: [1,2,3]
```
Thinking I will update param name `cb` to `callback` as it is more readable.

Need to update the documentation before I can merge this to the main branch, but its available under the `@beta` tag.

**Comment by AndersCan** [comment]
I've published v4.0.0 with worker state. Let me know if that doesn't work for you.

https://github.com/AndersCan/typed-web-workers#worker-with-local-state
